### PR TITLE
feat(sdk): add methods to find member/channel

### PR DESCRIPTION
**Pull request by qinezh**
Work item: https://msazure.visualstudio.com/Microsoft%20Teams%20Extensibility/_workitems/edit/15344675/
E2E TEST: N/A

### CODE REVIEW

File: packages/sdk/src/conversation/notification.ts [hunk 1]
```diff
@@ -65,6 +66,11 @@ export class Channel implements NotificationTarget {
    */
   public readonly info: ChannelInfo;
 
+  /**
+   * Team details.
+   */
+  public readonly team: TeamDetails;
```
Comment (swatDong) [inline packages/sdk/src/conversation/notification.ts:72]: Why the `TeamDetails` is part of channel, not team?

Comment (qinezh) [reply packages/sdk/src/conversation/notification.ts:72]: To support to get team details while finding channel, here's an example:
```ts
const channel = await teamsBot.notification.findChannel(
  c => c.info.name === "General" && c.team.name === "Contoso"
)
```

Comment (swatDong) [reply packages/sdk/src/conversation/notification.ts:72]: Then how about adding team details to both Team and Channel?
It's weird if team details can be only retrieved via channel.

Comment (swatDong) [reply packages/sdk/src/conversation/notification.ts:72]: Or, provide `findTeam` and `findChannel`. `findTeam` provides team details as input, while `findChannel` provides team details and channel info.

Comment (qinezh) [reply packages/sdk/src/conversation/notification.ts:72]: > Then how about adding team details to both Team and Channel? It's weird if team details can be only retrieved via channel.

I noticed there's only 'Member' and 'Channel' implements the 'NotificationTarget', do we have already implemented the 'Team'

Comment (swatDong) [reply packages/sdk/src/conversation/notification.ts:72]: `TeamsBotInstallation` may represent user, group, or team.

|Class|Type|Represent|
|-|-|-|
|TeamsBotInstallation|Person|1:1 chat|
|TeamsBotInstallation|Group|group chat|
|TeamsBotInstallation|Channel|team (general channel)|
|Channel|Channel|channel|
|Member|Person|1:1 chat|

My concern is - if providing team details, it should be provided for all team-related targets, including `TeamsBotInstallation(Channel)` and `Channel`.

File: packages/sdk/src/conversation/notification.ts [hunk 2]
```diff
@@ -433,4 +447,135 @@ export class NotificationBot {
...
+   * @param predicate find calls predicate once for each channel of the installation,
+   * until it finds one where predicate returns true. If such a channel is found, find
+   * immediately returns that channel. Otherwise, find returns undefined.
+   * @returns the first {@link Channel} where predicate is true, and undefined otherwise.
+   */
+  public async findChannel(
+    predicate: (channel: Channel) => Promise<boolean>
+  ): Promise<Channel | undefined> {
+    for (const target of await this.installations()) {
+      for (const channel of await target.channels()) {
```
Comment (swatDong) [inline packages/sdk/src/conversation/notification.ts:490]: Skip non-channel target(s) to reduce http calls.

File: packages/sdk/src/conversation/notification.ts [hunk 3]
```diff
@@ -362,6 +367,23 @@ export class TeamsBotInstallation implements NotificationTarget {
...
     return members;
   }
+
+  /**
+   * Get team details from this bot installation
+   *
+   * @returns the team details if bot is installed into a team, otherwise returns undefined.
+   */
+  public async getTeamDetails(): Promise<TeamDetails | undefined> {
+    let teamDetails: TeamDetails | undefined;
```
Comment (swatDong) [inline packages/sdk/src/conversation/notification.ts:377]: Skip if `type !== NotificationTargetType.Channel`?

File: packages/sdk/src/conversation/notification.ts [hunk 4]
```diff
@@ -433,4 +455,141 @@ export class NotificationBot {
...
+      (target.type === NotificationTargetType.Group && (scope & SearchScope.Group) !== 0) ||
+      (target.type === NotificationTargetType.Person && (scope & SearchScope.Person) !== 0)
+    );
+  }
+}
+
+/**
+ * The search scope when calling {@link NotificationBot.findMember} and {@link NotificationBot.findAllMembers}.
+ */
+export enum SearchScope {
```
Comment (swatDong) [inline packages/sdk/src/conversation/notification.ts:575]: Add documentation this enum is a flagged enum that can be combined via `|`.

File: packages/sdk/src/conversation/notification.ts (hunk 3 above)
Comment (qinezh) [reply packages/sdk/src/conversation/notification.ts:381]: good catch, fixed.

**Review by swatDong — APPROVED** [review]

**Review by dooriya — APPROVED** [review]

**Review by tecton — APPROVED** [review]

**Pull request merged**